ouroboros-consensus/test-storage/Test/Ouroboros/StateMachine.hs

```diff
@@ -0,0 +1,462 @@
+{-# LANGUAGE BangPatterns               #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveFunctor              #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures             #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE PartialTypeSignatures      #-}
+{-# LANGUAGE PolyKinds                  #-}
+-- {-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE TypeOperators              #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+
+module Test.Ouroboros.StateMachine (test) where
+
+import           Control.Monad.Free (Free)
+import           Control.Monad.Free as Free
+
+import           Data.ByteString hiding (hGet, hPut)
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.ByteString.Builder as BL
+import qualified Data.ByteString.Builder.Extra as BL
+import           Data.Either (isRight)
+import           Data.Kind(Type)
+import qualified Data.Map as Map
+import           Data.Map (Map, (!))
+import           Data.Maybe
+import           Control.Monad.State.Lazy
+import           Data.Typeable
+import           Data.Word (Word64)
+import           Data.TreeDiff (ToExpr)
+import           System.IO (IOMode (..), SeekMode (..))
+
+import           GHC.Generics (Generic, Generic1)
+import           Control.Exception (Exception(..), SomeException(..))
+import           Control.Monad.Except
+import           Control.Monad.Reader
+import           Control.Monad.Identity
+import           System.IO.Temp
+import           System.IO (IOMode(..))
+
+import           Ouroboros.Network.Sim
+
+import qualified System.Directory as Dir
+
+import           Test.Tasty
+                   (TestTree, defaultMain, testGroup, withResource)
+
+import           Test.Tasty.QuickCheck (testProperty, withMaxSuccess)
+
+import           Test.QuickCheck (arbitrary, Gen, ioProperty, Property, oneof, (===),
+                     elements, frequency, suchThat)
+import           Test.QuickCheck.Monadic (monadicIO, run)
+import           Test.StateMachine
+import qualified Test.StateMachine.Types.Rank2 as Rank2
+
+import           Ouroboros.Storage.FS.Sim
+import           Ouroboros.Storage.FS.IO
+import           Ouroboros.Storage.FS.Class
+
+import           Ouroboros.Network.MonadClass
+
+import           GHC.Stack
+
+type Fd = Int
+
+newtype Model (r :: Type -> Type) = Model
+    { unModel :: Opaque (MockFS, Int, Fd, Map Fd (MockHandle2, (Reference (Opaque (FsHandle IOFSE)) r)))}
```

Comment:
You can get nicer and more useful counterexamples by the `quickcheck-state-machine` library if you:

1. Avoid wrapping everything in `Opaque` (only use it for things that can't have a `Show` instance);
2. Use a record rather than a tuple (the `tree-diff` library has a way of compactly pretty printing diffs of records).